### PR TITLE
Ecc support

### PR DESCRIPTION
**Pull request by jschmidb**
This pr provides support for EC keys and ECDSA sign/verify ops. It also provides some bugfixes.

### CODE REVIEW

**Comment by ifranzki** [comment]
Travis shows build errors. Do you want to fix them first before we review? 

```
/home/travis/build/opencryptoki/libzpc/src/aes_key.c: In function ‘aes_key_blob_is_pkey_extractable’:
/home/travis/build/opencryptoki/libzpc/src/aes_key.c:1076:3: error: a label can only be part of a statement and a declaration is not a statement
   u16 kmf1 = ((struct aescipherkeytoken *)buf)->kmf1;
   ^~~
/home/travis/build/opencryptoki/libzpc/src/aes_key.c:1081:3: error: a label can only be part of a statement and a declaration is not a statement
   u64 attr = ((struct ep11keytoken *)buf)->attr;
   ^~~
```

File: src/ecc_key.c [hunk 1]
```diff
@@ -732,6 +788,36 @@ int zpc_ec_key_import_clear(struct zpc_ec_key *ec_key, const unsigned char *pubk
...
+
+			ep11_make_spki(ec_key->curve, ec_key->pub.pubkey, ec_key->pub.publen,
+					(unsigned char *)&temp, &temp_len);
+
+			ec_key->pub.spkilen = sizeof(ec_key->pub.spki);
+			rc = ep11_make_maced_spki(&ep11, temp, temp_len, ec_key->pub.spki,
+					&ec_key->pub.spkilen, target);
+
+			free_ep11_target_for_apqn(&ep11, target);
+			if (rc == 0) // Fixme: ist das �berall so drin?
```
Comment (vkanjur) [inline src/ecc_key.c:845]: Please remove comment

**Comment by jschmidb** [comment]
Yes, I will check ...

**Review by vkanjur — CHANGES_REQUESTED** [review]

**Comment by jschmidb** [comment]
OK, Travis is happy now.

File: src/ecc_key.c (hunk 1 above)
Comment (jschmidb) [reply src/ecc_key.c]: Done.

File: src/globals.c [hunk 2]
```diff
@@ -144,9 +147,13 @@ static void zpc_init(void)
 	assert(rc == 0);
 	if (load_cca_library(&cca, true) != 0)
 		DEBUG("loading CCA library failed");
-	else
+	else {
```
Comment (ifranzki) [inline src/globals.c:150]: Style: If the else has { ... } then the if should also have it.

File: src/globals.c [hunk 3]
```diff
@@ -158,9 +165,13 @@ static void zpc_init(void)
 	assert(rc == 0);
 	if (load_ep11_library(&ep11, true) != 0)
 		DEBUG("loading EP11 library failed");
-	else
+	else {
```
Comment (ifranzki) [inline src/globals.c:168]: Style: If the else has { ... } then the if should also have it.

File: src/globals.c [hunk 4]
```diff
@@ -158,9 +165,13 @@ static void zpc_init(void)
...
 		DEBUG("loaded EP11 library: %u.%u", ep11.version.major,
 	            ep11.version.minor);
+		aes_ep11 = 1;
+		if (ep11.version.major >= 3)
+			ecdsa_ep11 = 1;
```
Comment (ifranzki) [inline src/globals.c:173]: This is again just the EP11 host library version, but not the crypto card version.

File: src/zkey/pkey.c [hunk 5]
```diff
@@ -123,7 +167,39 @@ bool is_ep11_aes_key(const u8 *key, size_t key_size)
...
+	if (ep11key->version != EP11_STRUCT_MAGIC)
+		return false;
+
+	return true;
+}
+
+/**
+ * Check if the specified key is an EP11 ECC key token with header
+ * (TOKVER_EP11_ECC_WITH_HEADER). This means we have a 16-byte ep11kblob_header
+ * followed by a ep11keytoken struct with an own 32-byte header (session id).
```
Comment (ifranzki) [inline src/zkey/pkey.c:179]: the session Id is NOT a header, ist simply part of the EP11 key blob. So I would not state it that way....

File: src/zkey/pkey.h [hunk 6]
```diff
@@ -92,6 +92,65 @@ struct aescipherkeytoken {
...
+	/* followed by variable length q, max 133 bytes */
+} __packed;
+
+#define EP11_STRUCT_MAGIC        0x1234
+
+/*
+ * Internal used values for the version field of the key header.
+ * Should match to the enum pkey_key_type in pkey.h.
+ */
+#define TOKVER_EP11_AES  0x03  /* EP11 AES key blob (old style) */
```
Comment (ifranzki) [inline src/zkey/pkey.h:175]: I would add more space between the define and the value and align the values of the 3 defines.
But thats up to you....

File: src/zkey/pkey.h [hunk 7]
```diff
@@ -43,6 +43,31 @@ struct tokenheader {
...
+#define EC_PRIVLEN_P256                  32
+#define EC_PRIVLEN_P384                  48
+#define EC_PRIVLEN_P521                  66
+#define EC_PRIVLEN_ED25519               32
+#define EC_PRIVLEN_ED448                 57
+
+#define EC_BITLEN_P256                  256
+#define EC_BITLEN_P384                  384
+#define EC_BITLEN_P521                  521
+#define EC_BITLEN_ED25519               255
```
Comment (ifranzki) [inline src/zkey/pkey.h:62]: is this really 255 ? We use 256 at other places, but I am not sure what is the correct one.....

File: src/zkey/cca.c [hunk 8]
```diff
@@ -97,6 +110,17 @@ static int get_cca_version(struct cca_lib *cca, bool verbose)
 
 	util_assert(cca != NULL, "Internal error: cca is NULL");
 
+	if (cca->lib_csulcca == NULL) {
+		pr_verbose(verbose, "CCA host library not available.");
+		return -ELIBACC;
+	}
+
+	if (cca->dll_CSUACFV == NULL) {
```
Comment (ifranzki) [inline src/zkey/cca.c:118]: Can this really happen? Aren't you checking for all the functions being not NULL in load_cca_library() ? 
So when lib_csulcca is not NULL, then the functions should also all be there, no?

File: src/zkey/cca.c [hunk 9]
```diff
@@ -181,13 +220,16 @@ int load_cca_library(struct cca_lib *cca, bool verbose)
...
+	    cca->dll_CSNDPKG == NULL ||
+	    cca->dll_CSNDPKX == NULL ||
+	    cca->dll_CSNDPKI == NULL ||
+	    cca->dll_CSNDKTC == NULL) {
 		pr_verbose(verbose, "%s", dlerror());
 		DEBUG("The command requires the IBM CCA Host Libraries and "
 		      "Tools.\nFor the supported environments and downloads, "
 		      "see:\n%s", CCA_WEB_PAGE);
-		dlclose(cca->lib_csulcca);
-		cca->lib_csulcca = NULL;
```
Comment (ifranzki) [inline src/zkey/cca.c:190]: Why did you remove this? I would keep it that way (also see comment above).
If a certain function is optional, then don't check that one here, but check all the others.

Comment (ifranzki) [reply src/zkey/cca.c:190]: this would make all the checks for individual functions being NULL in the functions bleow.

File: src/zkey/cca.c [hunk 10]
```diff
@@ -587,3 +694,273 @@ int select_cca_adapter(struct cca_lib *cca, unsigned int card,
...
+		memcpy(rule_array_buf + *rule_array_count * CCA_KEYWORD_SIZE, "KEY-MGMT", CCA_KEYWORD_SIZE);
+		break;
+	case CCA_EC_CURVE_TYPE_EDWARDS:
+		memcpy(rule_array_buf + *rule_array_count * CCA_KEYWORD_SIZE, "SIG-ONLY", CCA_KEYWORD_SIZE);
+		break;
+	}
+	(*rule_array_count)++;
+
+	/* Add user-defined flags: the only flag that is currently recognized here,
+	 * is PKEY_KEYGEN_XPRT_AES : Allow export using an AES key. */
```
Comment (ifranzki) [inline src/zkey/cca.c:726]: Why do we need this? When is the EC key ever exported by an AES key ?

File: src/zkey/cca.c [hunk 11]
```diff
@@ -587,3 +694,273 @@ int select_cca_adapter(struct cca_lib *cca, unsigned int card,
...
+	} else {
+		memcpy(ecc_pub_token, pubtok->q, pubtok->q_len);
+		*p_ecc_pub_token_len = pubtok->q_len;
+	}
+
+	return 0;
+}
+
+int
+ec_key_clr2sec_cca(struct cca_lib *cca, int curve, unsigned int flags,
```
Comment (ifranzki) [inline src/zkey/cca.c:900]: int on same line ?

File: src/ecc_key.c [hunk 12]
```diff
@@ -0,0 +1,1222 @@
...
+		ep11hdr = (struct ep11kblob_header *)&buf;
+		ep11hdr->len = sizeof(struct ep11kblob_header) + key->seclen;
+		ep11hdr->version = PKEY_TYPE_EP11_ECC;
+		ep11hdr->bitlen = 0;
+		memcpy(buf + sizeof(struct ep11kblob_header), key->sec, key->seclen);
+		buflen = sizeof(struct ep11kblob_header) + key->seclen;
+
+		/* The ioctl wants the session id field at the beginning of the secure
+		 * key blob to be nulls, so clear it here in the temp buf. */
+		memset(buf + sizeof(struct ep11kblob_header), 0, 32);
```
Comment (ifranzki) [inline src/ecc_key.c]: hmm, really? What if the blob is a session-bound key? Then the session intentionally contains a session.
Note that for ep11 keys with header, you must leave the session field as is, but for ep11 keys where the session field is used as header, you must zeroize the session field.

File: src/zkey/ep11.c [hunk 13]
```diff
@@ -691,14 +743,14 @@ int ec_key_clr2sec_ep11(struct ep11_lib *ep11, unsigned int curve,
 
 	/* Add info to ep11 header (overlay over session ID field). This info
 	 * is later needed for reencipher. */
-	ep11hdr = (struct ep11kblob_header *)&blob;
-	ep11hdr->len = blob_len;
+	ep11hdr = (struct ep11kblob_header *)secure_key;
```
Comment (ifranzki) [inline src/zkey/ep11.c:746]: The comment above is now wrong (overlay over session ID field

File: src/ecc_key.c [hunk 14]
```diff
@@ -943,9 +1033,30 @@ int zpc_ec_key_reencipher(struct zpc_ec_key *ec_key, unsigned int method)
 			if (rc)
 				continue;
 
+			/* Note that the secure key is a TOKVER_EP11_ECC_WITH_HEADER and has a
+			 * 16-byte ep11kblob_header prepended before the actual secure key blob.
+			 * For reencipher we have to skip this prepended hdr and provide the
+			 * ep11kblob_header info as an overlay over the session field at
+			 * the beginning of the secure key. So at this point we assume that
+			 * the key's session id field does not contain any info! */
+			ep11hdr = (struct ep11kblob_header *)((unsigned char *)reenc.sec +
```
Comment (ifranzki) [inline src/ecc_key.c:1073]: Please check if you really overlay the session field here. 
I would strongly suggest to ALWAYS use the EP11 key with header IN FRONT, and not the overlayed one.

File: src/ecc_key.c [hunk 15]
```diff
@@ -1115,27 +1227,20 @@ int ec_key_sec2prot(struct zpc_ec_key *ec_key, enum ec_key_sec sec)
...
-		memcpy(buf + sizeof(struct ep11kblob_header), key->sec, key->seclen);
-		buflen = sizeof(struct ep11kblob_header) + key->seclen;
-
+		keybuf = key->sec;
+		keybuf_len = key->seclen + sizeof(struct ep11kblob_header);
 		/* The ioctl wants the session id field at the beginning of the secure
-		 * key blob to be nulls, so clear it here in the temp buf. */
-		memset(buf + sizeof(struct ep11kblob_header), 0, 32);
+		 * key blob to be nulls, so clear it here. */
+		memset(key->sec + sizeof(struct ep11kblob_header), 0, 32);
```
Comment (ifranzki) [inline src/ecc_key.c:1266]: Please check if this is correct in regards of the session

File: src/ecc_key.c [hunk 16]
```diff
@@ -1425,3 +1449,28 @@ static int ec_key_spki_has_valid_mkvp(const struct zpc_ec_key *ec_key,
...
+{
+	const unsigned char *mkvp;
+	unsigned int mkvp_len;
+
+	if (ec_key->mkvp_set == 0)
+		return 1; /* cannot judge */
+
+	if (ec_key->type == ZPC_EC_KEY_TYPE_CCA) {
+		mkvp = ((struct ccakeytoken *)buf)->mkvp;
+		mkvp_len = 8;
```
Comment (ifranzki) [inline src/ecc_key.c:1470]: Maybe use defines in the corresponding header ?
Same below for EP11

File: src/aes_key.c [hunk 17]
```diff
@@ -1056,3 +1063,31 @@ static int aes_key_blob_has_valid_mkvp(struct zpc_aes_key *aes_key, const unsign
...
+{
+	u16 kmf1;
+	u64 attr;
+
+	switch (aes_key->type) {
+	case ZPC_AES_KEY_TYPE_CCA_DATA:
+		/* No check possible. The flags field in struct aesdatakeytoken
+		 * does not contain a CCA_XPRTCPAC indication. */
+		return 1;
+		break;
```
Comment (ifranzki) [inline src/aes_key.c:1077]: no break needed after return

File: src/globals.c (hunk 2 above)
Comment (jschmidb) [reply src/globals.c:150]: Done.

File: src/globals.c (hunk 3 above)
Comment (jschmidb) [reply src/globals.c:168]: Done.

File: src/zkey/pkey.c (hunk 5 above)
Comment (jschmidb) [reply src/zkey/pkey.c:179]: Done.

File: src/zkey/pkey.c [hunk 18]
```diff
@@ -123,7 +167,39 @@ bool is_ep11_aes_key(const u8 *key, size_t key_size)
...
+ * We assume that the blob does not contain a filled out ep11keytoken header
+ * in the session field.
```
[the remark this answers is not in the document]
Comment (jschmidb) [reply src/zkey/pkey.c:181]: OK, done.

File: src/zkey/pkey.h (hunk 6 above)
Comment (jschmidb) [reply src/zkey/pkey.h:175]: Done.

File: src/zkey/pkey.h (hunk 7 above)
Comment (jschmidb) [reply src/zkey/pkey.h:62]: Yes indeed. I have it from the CCA 7.3 book, chapter ECC key token, Table 345. Supported Edwards elliptic curves: Size of prime p in bits (key length): 

> 255    1.3.101.112     id-Ed25519    Curve25519

File: src/zkey/cca.c (hunk 11 above)
Comment (jschmidb) [reply src/zkey/cca.c:900]: Done. Yes, in the existing code the return type is often coded in a separate line, while I'm usually prefer to have it on the same line with the function name.

File: src/zkey/ep11.c [hunk 19]
```diff
@@ -140,18 +152,32 @@ int load_ep11_library(struct ep11_lib *ep11, bool verbose)
...
+	    ep11->dll_m_EncryptSingle == NULL ||
+	    ep11->dll_m_GenerateKeyPair == NULL ||
+	    ep11->dll_m_GenerateKey == NULL ||
+	    ep11->dll_m_GetAttributeValue == NULL) {
 		pr_verbose(verbose, "%s", dlerror());
 		DEBUG("The command requires the IBM Z Enterprise PKCS #11 "
 		      "(EP11) Support Program (EP11 host library).\n"
 		      "For the supported environments and downloads, see:\n%s",
 		      EP11_WEB_PAGE);
+		all_entry_points_loaded = 0;
```
[the remark this answers is not in the document]
Comment (jschmidb) [reply src/zkey/ep11.c:231]: This is exactly the question. I can't say if we can assume all entry points being available when the lib load was ok. These checks want to make the code more robust.

File: include/zpc/ecc_key.h [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+
+/**
+ * Import an EC secure-key.
+ * \param[in,out] key EC key
+ * \param[in] seckey EC secure-key
+ * \param[in] seckeylen EC key secure-length [bytes]
+ * \return 0 on success. Otherwise, a non-zero error code is returned.
+ */
+__attribute__((visibility("default")))
+int zpc_ec_key_import(struct zpc_ec_key *key, const unsigned char *seckey,
```
[the remark this answers is not in the document]
Comment (jschmidb) [reply include/zpc/ecc_key.h:108]: Yes. Comment added.

File: src/ecc_key.c (hunk 16 above)
Comment (jschmidb) [reply src/ecc_key.c:1470]: OK, introduced 2 constants MKVP_LEN_CCA and MKVP_LEN_EP11.
Done.

File: src/zkey/ep11.c (hunk 13 above)
Comment (jschmidb) [reply src/zkey/ep11.c:740]: Corrected.

File: src/zkey/ep11.c [hunk 21]
```diff
@@ -402,3 +446,421 @@ int reencipher_ep11_key(struct ep11_lib *ep11, target_t target,
...
+						ep11_pin_blob, ep11_pin_blob_len,
+						blob, &blob_len,
+						spki, &spki_len, target);
+	if (rv != CKR_OK) {
+		DEBUG("Error m_GenerateKeyPair, rv = 0x%lx", rv);
+		rc = EIO;
+		goto ret;
+	}
+
+	/* Add info to ep11 header (overlay over session ID field). This info
```
Comment (ifranzki) [inline src/zkey/ep11.c]: here you still mention 'overlay' ? I guess thats no longer true, right?

Comment (ifranzki) [reply src/zkey/ep11.c]: hm, looks like you are still overlaying here, so this needs to be changed to a blob with a prepended header!

File: src/ecc_key.c (hunk 14 above)
Comment (ifranzki) [reply src/ecc_key.c]: I guess this still needs to be fixed to not use an overlay of the session field?

File: src/ecc_key.c [hunk 22]
```diff
@@ -1067,17 +1066,9 @@ int zpc_ec_key_reencipher(struct zpc_ec_key *ec_key, unsigned int method)
...
-			ep11hdr = (struct ep11kblob_header *)((unsigned char *)reenc.sec +
-					sizeof(struct ep11kblob_header));
-			ep11hdr->len = ec_key->cur.seclen - sizeof(struct ep11kblob_header);
-			ep11hdr->version = PKEY_TYPE_EP11_ECC;
-			ep11hdr->bitlen = 0;
-
+			 * key blob directly. */
 			rc = reencipher_ep11_key(&ep11, target, ec_key->apqns[i].card,
-					ec_key->apqns[i].domain, (unsigned char *)ep11hdr,
+					ec_key->apqns[i].domain, reenc.sec + sizeof(struct ep11kblob_header),
```
Comment (ifranzki) [inline src/ecc_key.c:1076]: Ah OK, here you fix that, so please forget my previous comment on that place on an earlier commit.
You could incorporate that change for EC key into the earlier commit though, where you introduce the code freshly....

File: src/aes_key.c [hunk 23]
```diff
@@ -1099,3 +1205,16 @@ static int aes_key_blob_is_pkey_extractable(struct zpc_aes_key *aes_key, const u
 
 	return 0;
 }
+
+static void aes_key_add_ep11_header(struct zpc_aes_key *aes_key)
+{
+	struct ep11kblob_header *ep11hdr;
+
+	memset(aes_key->cur.sec, 0, sizeof(struct ep11kblob_header));
+	memmove(aes_key->cur.sec + sizeof(struct ep11kblob_header), aes_key->cur.sec, aes_key->cur.seclen);
```
Comment (ifranzki) [inline src/aes_key.c:1153]: Are you sure that the size of the buffer is large enough here?

File: src/aes_key.c [hunk 24]
```diff
@@ -1099,3 +1205,16 @@ static int aes_key_blob_is_pkey_extractable(struct zpc_aes_key *aes_key, const u
...
+
+	ep11hdr = (struct ep11kblob_header *)aes_key->cur.sec;
+	ep11hdr->len = sizeof(struct ep11kblob_header) + aes_key->cur.seclen;
+	ep11hdr->version = TOKVER_EP11_AES_WITH_HEADER;
+	aes_key->cur.seclen = ep11hdr->len;
```
Comment (ifranzki) [inline src/aes_key.c:1158]: You should zero the session field in the blob, now that you have added the header. Otherwise the session would specify a non-existing session id and it would probably fail.

File: src/aes_key.c [hunk 25]
```diff
@@ -488,6 +489,9 @@ zpc_aes_key_import_clear(struct zpc_aes_key *aes_key, const unsigned char *key)
 		goto ret;
 	}
 
+	if (aes_key->type == ZPC_AES_KEY_TYPE_EP11)
```
Comment (ifranzki) [inline src/aes_key.c:490]: Who defines the type of the key? Shouldn't it in case of EP11 always be the one with header prepended? Would you ever create one with overlayed header?

File: src/aes_key.c [hunk 26]
```diff
@@ -190,8 +188,8 @@ zpc_aes_key_set_type(struct zpc_aes_key *aes_key, int type)
 		aes_key->napqns = 0;
 		aes_key->apqns_set = 0;
 
-		rc = __aes_key_alloc_apqns_from_mkvp(&(aes_key->apqns),
-		    &(aes_key->napqns), aes_key->mkvp, type);
+		rc = alloc_apqns_from_mkvp(pkeyfd, &(aes_key->apqns), &(aes_key->napqns),
```
Comment (ifranzki) [inline src/aes_key.c:191]: I noticed that originally all internal helper functions were named __xxxx with 2 leading underscores. You seem to have given up this naming scheme..... Is this intentionally?

File: src/globals.c [hunk 27]
```diff
@@ -144,9 +147,13 @@ static void zpc_init(void)
 	assert(rc == 0);
 	if (load_cca_library(&cca, true) != 0)
 		DEBUG("loading CCA library failed");
-	else
+	else {
 		DEBUG("loaded CCA library: ver %u, rel %u, mod %u",
 	            cca.version.ver, cca.version.rel, cca.version.mod);
+		aes_cca = 1;
+		if (cca.version.ver >= 7)
```
[the remark this answers is not in the document]
Comment (jschmidb) [reply src/globals.c:154]: Fixed: see commit "ECC: check for min card versions"

File: src/ecc_key.c (hunk 12 above)
Comment (jschmidb) [reply src/ecc_key.c]: Fixed. Code has been reworked.

File: src/ecc_key.c (hunk 14 above)
Comment (jschmidb) [reply src/ecc_key.c]: Fixed. Code is reworked.

File: src/zkey/ep11.c [hunk 28]
```diff
@@ -402,3 +446,421 @@ int reencipher_ep11_key(struct ep11_lib *ep11, target_t target,
...
+	};
+	int rc = -EIO;
+
+	if (ep11->dll_m_GetAttributeValue == NULL) {
+		DEBUG("EP11 host lib function m_GetAttributeValue not "
+			"available but required for EP11 EC extract public key.");
+		return -ELIBACC;
+	}
+
+	rv = ep11->dll_m_GetAttributeValue(secure_key, seclen, templ, 1, target);
```
[the remark this answers is not in the document]
Comment (jschmidb) [reply src/zkey/ep11.c]: Code is reworked: all keys are now "keys with header" and the call looks like:
```
/* Note that the secure key is a TOKVER_EP11_ECC_WITH_HEADER and has a
 * 16-byte ep11kblob_header prepended before the actual secure key blob.
 * This header must not be passed to the ep11 library. */
rv = ep11->dll_m_GetAttributeValue(secure_key + sizeof(struct ep11kblob_header),
					seclen - sizeof(struct ep11kblob_header),
					templ, 1, target);
```
For the ep11 lib call I just have to skip the header.

File: src/zkey/ep11.c (hunk 21 above)
Comment (jschmidb) [reply src/zkey/ep11.c]: Code is reworked. No overlay anymore except of one place: in aes_key.c, routine aes_key_sec2prot_with_header the ioctl wants a TOKEN_VERSION_EP11_AES (i.e. no header). But there I'm saving the original data in the first 16 bytes and I'm restoring them after the ioctl.

File: src/aes_key.c (hunk 23 above)
Comment (jschmidb) [reply src/aes_key.c:1153]: Yes, the sec buffer has MAX_AESKEYBLOBSIZE = 512 bytes, so plenty of space.

File: src/zkey/ep11.c (hunk 28 above)
[the remark this answers is not in the document]
Comment (ifranzki) [reply src/zkey/ep11.c]: Right, the call looks good now.

I was wondering if the EP11 host lib supports a GetAttribueValues with  CKA_EC_POINT at all. Its not documented that it would support so (and I can't see that in the EP11 code being implemented either). So I assume that this call will actually never work.....

Newer host libs support GetAttribueValues with CKA_PUBLIC_KEY_INFO to get the SPKI....

File: src/aes_key.c (hunk 23 above)
Comment (ifranzki) [reply src/aes_key.c:1153]: I meant to add a check into the code, and not just trust that it is enough :-)

File: src/zkey/ep11.c (hunk 21 above)
Comment (ifranzki) [reply src/zkey/ep11.c]: > ioctl wants a TOKEN_VERSION_EP11_AES

Isn't there a newer ioctl command available that works on tokens with external header?

File: src/aes_key.c (hunk 25 above)
Comment (jschmidb) [reply src/aes_key.c:490]: Yes, this is a libzpc constant that applies for aes ep11 keys in general (in contrast to CCA). EP11 type keys always get the header:
# define ZPC_AES_KEY_TYPE_CCA_DATA     1
# define ZPC_AES_KEY_TYPE_CCA_CIPHER   2
# define ZPC_AES_KEY_TYPE_EP11         3

File: src/aes_key.c (hunk 26 above)
Comment (jschmidb) [reply src/aes_key.c:191]: Yes, I didn't always continue with the previous naming scheme. Especially for this routine the decision was: previously there was the __aes_key_alloc_apqns_from_mkvp. For ECC I first introduced a corresponding __ec_key_alloc_apqns_from_mkvp. But this routine was completely identical to the aes routine. So I consolidated the two into alloc_apqns_from_mkvp in pkey.c where no other routine has the "__" prefix.

File: src/aes_key.c (hunk 24 above)
Comment (jschmidb) [reply src/aes_key.c:1158]: Fixed.

File: src/aes_key.c (hunk 23 above)
Comment (jschmidb) [reply src/aes_key.c:1153]: OK fixed ;-)

File: src/zkey/ep11.c (hunk 21 above)
Comment (jschmidb) [reply src/zkey/ep11.c]: Unfortunately not. But I noted the point for the future.

File: src/zkey/cca.c (hunk 8 above)
Comment (jschmidb) [reply src/zkey/cca.c:118]: The idea here was to in fact allow single functions to be NULL and always check the function ptr "at use". I agree that in the past it was like you say "lib_csulcca is not NULL, then the functions should also all be there". But this may change as we introduce more functions from newer libs, but customers still use older ones.

File: src/zkey/cca.c (hunk 9 above)
Comment (jschmidb) [reply src/zkey/cca.c:190]: see comment above.

File: src/zkey/ep11.c (hunk 28 above)
[the remark this answers is not in the document]
Comment (jschmidb) [reply src/zkey/ep11.c]: Good catch. Reworked to use CKA_IBM_MACED_PUBLIC_KEY_INFO, which however is only supported with ep11 hostlib v4.

File: src/ecc_key.c [hunk 29]
```diff
@@ -0,0 +1,1528 @@
...
+		}
+		ec_key->pubkey_set = 1;
+		rv = pthread_mutex_lock(&ep11lock);
+		assert(rv == 0);
+		for (i = 0; i < ec_key->napqns; i++) {
+			rc = get_ep11_target_for_apqn(&ep11, ec_key->apqns[i].card,
+					ec_key->apqns[i].domain, &target, true);
+			if (rc)
+				continue;
+			rc = ec_key_extract_public_ep11(&ep11, ec_key->curve,
```
Comment (ifranzki) [inline src/ecc_key.c:685]: Hm, so in ec_key_extract_public_ep11 you obtain the MACed SPKI, but you only use the EC-point at this time. 
First question: Why do you obtain the MACed SPKI then, you only really need the raw SPKI for that.
Second question: Don't you also want to have the MACed SPKI for such a key and keep in in the key object? So instead of extracting just the EC point, pass back the whole MACed-SPKI?

Comment (jschmidb) [reply src/ecc_key.c:687]: Good point! Code changed.

**Review by ifranzki — APPROVED** [review]
Looks good, as far as I can tell.

**Review by vkanjur — APPROVED** [review]
